src/libexpr/primops/regex.cc

```diff
@@ -0,0 +1,360 @@
+
+      ```nix
+      builtins.match "abc" "abc"
+      ```
+
+      Evaluates to `[ ]`.
+
+      ```nix
+      builtins.match "a(b)(c)" "abc"
+      ```
+
+      Evaluates to `[ "b" "c" ]`.
+
+      ```nix
+      builtins.match "[[:space:]]+([[:upper:]]+)[[:space:]]+" "  FOO   "
+      ```
+
+      Evaluates to `[ "FOO" ]`.
+    )s",
+    .fun = prim_match,
+});
+
+/* Split a string with a regular expression, and return a list of the
+   non-matching parts interleaved by the lists of the matching groups. */
+void prim_split(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto re = state.forceStringNoCtx(*args[0], pos);
+
+    try {
+        auto& regex = state.regexCache->get(re);
+
+        // We're going to use this regex many times, JIT it.
+        regex.compile();
+
+        PathSet context;
+        const auto str = state.forceString(*args[1], context, pos);
+
+
+        PCRE::MatchData match(regex);
+        int rc = regex.match(str, match);
+
+        if (rc == PCRE2_ERROR_NOMATCH) {
+            state.mkList(v, 1);
+            v.listElems()[0] = args[1];
+            return;
+        }
+        ValueVector result;
+        result.reserve(3);
+        size_t lastmatchend = 0;
+        size_t newmatchstart = 0;
+        while (rc != PCRE2_ERROR_NOMATCH && newmatchstart <= str.length()) {
+            // Add a string for non-matched characters.
+            Value * prefix = state.allocValue();
+            prefix->mkString(str.substr(lastmatchend, match.startPos() - lastmatchend));
+            result.push_back(prefix);
+
+            // Add a list for matched substrings.
+            auto elem = state.allocValue();
+            result.push_back(elem);
+
+            extract_matches(state, match, *elem);
+
+            lastmatchend = match.startPos() + match[0]->length();
+            newmatchstart = lastmatchend + (match[0]->length() == 0 ? 1 : 0);
+            if (newmatchstart <= str.length())
+                rc = regex.match(str, match, newmatchstart);
+        }
+
+        // Add a string for non-matched suffix characters.
+        Value * rest = state.allocValue();
+        rest->mkString(str.substr(lastmatchend, -1));
+        result.push_back(rest);
+
+        state.mkList(v, result.size());
+        for (size_t n = 0; n < result.size(); ++n)
+            v.listElems()[n] = result[n];
+
+    } catch (RegexError & e) {
+        state.debugThrowLastTrace(EvalError({
+            .msg = hintfmt("error while evaluating regex '%s': ", re, e.what()),
+            .errPos = state.positions[pos]
+        }));
+    }
+}
+
+static RegisterPrimOp primop_split({
+    .name = "__split",
+    .args = {"regex", "str"},
+    .doc = R"s(
+      Returns a list composed of non matched strings interleaved with the
+      lists of the [extended POSIX regular
+      expression](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_04)
+      *regex* matches of *str*. Each item in the lists of matched
```

Comment:
PCRE is not POSIX ERE.

Also worth mentioning that PCRE supports arbitrary look-around, which can cause (maybe accidental) exponential run time. Maybe we should limit the input to a reasonable subset.

Reply:
So it should point to https://www.pcre.org/original/doc/html/pcrepattern.html.